Task 2 implement productlist page

VladislawL opened the pull request.

BuisnessMonkeyy committed:
Added sort, PhoneService, master and sortLink tags, stylized pages

BuisnessMonkeyy committed:
Added search and pagination, implemented DefaultPhoneService

BuisnessMonkeyy committed:
Added CartItem, implemented Cart, HttpSessionCartService and AjaxCartController

BuisnessMonkeyy committed:
Added AjaxExceptionHadler and CartItemValidationException

BuisnessMonkeyy committed:
Added tests for HttpSessionCartService, StockValidator and AjaxCartController

BuisnessMonkeyy committed:
Changed ContextConfiguration in AjaxCartControllerTest

AndreikaInactive commented on core/src/main/java/com/es/core/cart/CartItem.java:
```diff
@@ -0,0 +1,57 @@
+public class CartItem implements Serializable {
+
+    @NotNull
+    private long phoneId;
```
не будет проблем с 0 по дефолту?

VladislawL replied:
По идее не должно быть. Этот объект будет создан только по информации, что пришла от клиента. А если будет какое-то несуществующее id или тот же 0, то телефон просто не будет найден.

AndreikaInactive replied:
в данном случае мб и ок, но вообще не оч хорошо что мы не можем точно понять - новый это объект или уже сохранённый в бд

AndreikaInactive commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,78 @@
     @Override
     public void addPhone(Long phoneId, Long quantity) {
-        throw new UnsupportedOperationException("TODO");
+        if (cart.getCartItems().stream().anyMatch(cartItem -> cartItem.getPhoneId().equals(phoneId))) {
```
почему не использовать findCartItem и здесь?

AndreikaInactive commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,78 @@
+        cart.setSubTotalPrice(calculateSubTotalPrice());
+    }
+
+    private BigDecimal calculateSubTotalPrice() {
```
вынеси логику подсчёта в отдельный сервис

AndreikaInactive replied:
?

VladislawL replied:
Стоп, так у меня этого уже нет. Я вынес подсчёт прайса в калькулятор

AndreikaInactive commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,78 @@
+
+        for (CartItem cartItem : cartItems) {
+            Phone phone = phoneService.getPhoneById(cartItem.getPhoneId());
+            subTotalPrice = subTotalPrice.add(phone.getPrice().multiply(BigDecimal.valueOf(cartItem.getQuantity())));
```
можно сделать через стримы

AndreikaInactive replied:
а почему зарезолвал?

VladislawL replied:
И тут пофиксано,

AndreikaInactive commented on core/src/main/java/com/es/core/dao/JdbcPhoneDao.java:
```diff
@@ -48,6 +57,8 @@
 
     private final String JOIN_PHONE_AND_COLOR_QUERY = "insert into phone2color (phoneId, colorId) values (:phoneId, :colorId)";
 
+    private final String FIND_ALL_PHONES_WITH_OFFSET_AND_LIMIT_QUERY = "select * from phones offset :offset limit :limit";
```
static

AndreikaInactive commented on core/src/main/java/com/es/core/dao/JdbcStockDao.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public long getStock(long phoneId) {
+        Map<String, Object> parameters = new HashMap<>();
```
логика подсчёта стока может быть значительно сложнее - лучше возвращай сток, а итоговое колво считай в сервисе

AndreikaInactive commented on core/src/main/java/com/es/core/model/phone/SortField.java:
```diff
@@ -0,0 +1,19 @@
+package com.es.core.model.phone;
+
+public enum SortField {
```
возможные сортировки лучше держать в конфиге, т.к потенциально будут изменяться

AndreikaInactive commented on core/src/main/java/com/es/core/services/DefaultPhoneService.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public List<Phone> getPhonePage(int page, String query, SortField sortField, SortOrder sortOrder) {
+        int limit = 10;
```
а лимит должен быть захардкодан или передаваться парамом? как минимум в конфиг вынести

AndreikaInactive commented on core/src/main/java/com/es/core/services/DefaultPhoneService.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public int getPagesCount(String searchQuery) {
+        return phoneDao.countPhonesWhereBrandAndModelLikeSearchQuery(searchQuery) / 10;
```
и здесь непонятно

AndreikaInactive commented on core/src/main/java/com/es/core/services/DefaultPhoneService.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public Phone getPhoneById(long id) {
+        return phoneDao.get(id).get();
```
а если не будет в базе?

AndreikaInactive commented on core/src/main/java/com/es/core/validators/StockValidator.java:
```diff
@@ -0,0 +1,38 @@
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class StockValidator implements ConstraintValidator<QuantityConstraint, Object> {
```
во первых непонятно почему относится именно к QuantityConstraint , во вторых - выглядит как типичный оверинжиниринг, понятней явно проверять перед изменением айтема

AndreikaInactive commented on pom.xml:
```diff
@@ -51,13 +51,25 @@
         <scope>test</scope>
       </dependency>
 
+      <dependency>
```
а почему в помы добавляются одни и те же депенденси?

VladislawL replied:
Я не нашёл повторяющихся зависимостей

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/advicer/AjaxExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+    @ExceptionHandler(CartItemValidationException.class)
+    public ResponseEntity<Map<String, Object>> handleCartItemValidationException(Exception e) {
+        Map<String, Object> errorMessage = new HashMap<>();
```
сделай отдельный бин под эрроры

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/AjaxCartController.java:
```diff
@@ -14,7 +24,17 @@
+            cartService.addPhone(cartItem.getPhoneId(), cartItem.getQuantity());
+            Cart cart = cartService.getCart();
+            Map<String, Object> miniCart = new HashMap<>();
+            miniCart.put("subTotalPrice", cart.getSubTotalPrice().toString());
```
сделай отдельный бин под миникарт, добавь для неё отдельный метод, не надо форматировать через toString

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/AjaxCartController.java:
```diff
@@ -14,7 +24,17 @@
+            miniCart.put("itemsNumber", cart.getCartItems().size());
+            return ResponseEntity.ok(miniCart);
+        } else {
+            throw new CartItemValidationException(errors.getAllErrors().get(0).getDefaultMessage());
```
а остальные зачем терять? передавай весь errors в эксепшен

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/pages/ProductListPageController.java:
```diff
@@ -1,23 +1,42 @@
+            Model model) {
+        int pagesCount = phoneService.getPagesCount(query);
+
+        model.addAttribute("phones", phoneService.getPhonePage(page, query, sortField, sortOrder));
```
если возвращает пейдж - по идее там сразу же должна быть и дата по пагинации

VladislawL replied:
Вот тут я не понял. Нужно прям все найденные телефоны сетать в атрибут?

AndreikaInactive replied:
сделай pagination data бин и туда сетай телефоны, плюс pagesCount, сортировку, страницу итд

AndreikaInactive commented on web/src/test/resources/context/applicationContext.xml:
```diff
@@ -0,0 +1,16 @@
+                           http://www.springframework.org/schema/context http://www.springframework.org/schema/context/spring-context.xsd
+                           http://www.springframework.org/schema/jdbc http://www.springframework.org/schema/jdbc/spring-jdbc.xsd http://www.springframework.org/schema/tx http://www.springframework.org/schema/tx/spring-tx.xsd">
+
+    <import resource="applicationContext-core.xml" />
```
а это всё зачем?

VladislawL replied:
По другому не работало, тесты всё время падали с FileNotFoundException. Я думал так и должно быть. Тогда странно получается, попробую исправить чуть что отпишусь

AndreikaInactive replied:
это надо удалить, ты скорей тесты не так запускаешь

AndreikaInactive commented on web/src/test/resources/config/application.properties:
```diff
@@ -0,0 +1,7 @@
+db.driver=org.h2.Driver
```
достаточно того что в коре

VladislawL replied:
Context в тестовой папке я удалил, а вот без этого файла падали тесты, где нужна была бд

AndreikaInactive replied:
так а что здесь? в вебе и тестах не надо дупликатов, достаточно того что в коре лежит, ты же его подключаешь в тесты

VladislawL replied:
Ну да я подключаю, но тесты почему то падают с FileNotFoundException и указывается файл с test-data.sql, хота например тот же schema.sql загружается. Тогда попробую что нибудь придумать, может я где-то что то не так делал

AndreikaInactive commented on core/src/main/java/com/es/core/dao/JdbcPhoneDao.java:
```diff
@@ -91,6 +115,22 @@ private void add(final Phone phone) {
     }
 
+    private String getBrandAndModelLikeSearchQueryCondition(String searchQuery) {
+        StringBuilder result = new StringBuilder("lower(brand) like '%" + searchQuery.toLowerCase() +
```
уязвимо для sql injection - парметры надо сетать только через ? :name

AndreikaInactive commented on core/src/main/java/com/es/core/dao/JdbcPhoneDao.java:
```diff
@@ -66,7 +77,20 @@ public void save(final Phone phone) {
+
+    public List<Phone> findOrderedPhoneListBySearchQuery(int offset, int limit, String searchQuery, SortOrder sortOrder, SortField sortField) {
+        return namedParameterJdbcTemplate.query("select * from phones join stocks on id = phoneId where (" + getBrandAndModelLikeSearchQueryCondition(searchQuery) +
+                ") and stock > 0 and price > 0 order by " + sortField.getAttribute() + " " + sortOrder.name() + " offset " + offset + " limit " + limit, phoneRowMapper);
```
а если в полях будет null?

VladislawL replied:
А что в таком случае стоит сделать, бросить ошибку или использовать какое-нибудь дефолтное значение?

AndreikaInactive commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,78 @@
     public void addPhone(Long phoneId, Long quantity) {
-        throw new UnsupportedOperationException("TODO");
+        if (cart.getCartItems().stream().anyMatch(cartItem -> cartItem.getPhoneId().equals(phoneId))) {
+            Map<Long, Long> cartItem = new HashMap<>();
```
попробуй выделить методы add/update phone и закинуть всю в них логику по валидации итд и переиспользовать в публичных addPhone и update

VladislawL replied:
Я так понимаю будет без валидации спринга. Даже без @ Positive или @ NotNull

VladislawL replied:
Positive и NotNull я оставил, но со стороны выглядит как-то не удобно, что при большом quantity я кидаю ошибку, а тут через errors чекаю. Может стоит всюду кидать ошибку, а в контроллере её хэндлить и упаковывать их в другую ошибку, которую в свою очередь перехватит ExceptionHandler.

AndreikaInactive replied:
аннотации хэндлать выше, на уровне валидатора формы - там да, показывать на ui, а здесь разве что ассертнуть что qty валидный инт

BuisnessMonkeyy committed:
Moved price calculations to PriceCalculator service
Fixed HttpSessionCartService, JdbcPhoneDao getPhone() and QuantityValidation
Added PriceCalculator, MiniCart, sortFields.properties and webApplication.properties

AndreikaInactive commented on web/src/main/webapp/WEB-INF/tags/sortLink.tag:
```diff
@@ -0,0 +1,8 @@
+<%@ attribute name="order" required="true" %>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+
+<a href="<c:url value="/productList?query=${param.query}&sortField=${field}&sortOrder=${order}&page=1"/>" />
```
используя c url с парамами вместо мануальной конструкции

AndreikaInactive commented on web/src/main/webapp/WEB-INF/tags/pagination.tag:
```diff
@@ -0,0 +1,43 @@
+
+<c:choose>
+
+    <c:when test="${currentPage - 6 < 0}">
```
что то вообще жесткий хардкод пошёл ) попробуй работать с параметрами pagination data объекта из контроллера - с текущей страницей, колвом результатов на страницу, общим колвом результатов

AndreikaInactive commented on web/src/main/webapp/WEB-INF/tags/master.tag:
```diff
@@ -0,0 +1,43 @@
+            <div class="mini-cart">
+                <i class="fas fa-shopping-cart"></i>
+                <span id="items-number">${cart.cartItems.size()} items</span>
+                <span id="subtotal-price">$${cart.subTotalPrice}</span>
```
formatPrice тег какой используй из fn

AndreikaInactive commented on web/src/main/webapp/WEB-INF/tags/pageLink.tag:
```diff
@@ -0,0 +1,8 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+
+<a href="<c:url value="/productList?query=${param.query}&sortField=${param.sortField}&sortOrder=${param.sortOrder}&page=${pageNumber}"/>" class="page-link" />
+    ${pageText}
```
текст обычно надёжнее рисовать через c:out - ескейпит всякие xss-сы

AndreikaInactive commented on web/src/main/webapp/WEB-INF/pages/productList.jsp:
```diff
@@ -1,20 +1,44 @@
+    <thead class="thead-light">
+    <tr>
+      <th>Image</th>
+      <th>
```
можно просто возвращать список полей для сортировки из контроллера и хэндлать в цикле

AndreikaInactive replied:
сделай модель аттрибут в контроллере и оттуда все филды передавай

AndreikaInactive replied:
м?

VladislawL replied:
Так ну я вроде сделал. Но есть вопросы. Всё ли хорошо с этим циклом, ну в том плане что мы же потом в другом цикле строки для телефонов формируем, и мне кроме как if 'ов в голову ничего не приходит + если в проперти файле атрибуты поменять местами, то уже колонки будут не соответствовать. В общем то будет ли это проблемой в будущем?

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/converter/StringToSortOrder.java:
```diff
@@ -0,0 +1,16 @@
+    @Override
+    public SortOrder convert(String sortOrder) {
+        try {
+            return Enum.valueOf(SortOrder.class, sortOrder);
```
думаю спринг и сам умеет это конвертить в контроллере, чекни

VladislawL replied:
Да, спринг и сам сможет это сделать. Но если в enum'е не найдётся константа, то будет брошена ошибка и я бы хотел такой случай перехватить и использовать какое-нибудь дефолтное значение

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/converter/StringToSortFieldConverter.java:
```diff
@@ -0,0 +1,16 @@
+    @Override
+    public SortField convert(String sortField) {
+        try {
+            return Enum.valueOf(SortField.class, sortField);
```
см ниже

AndreikaInactive commented on core/src/main/java/com/es/core/cart/CartItemValidationException.java:
```diff
@@ -1,7 +1,24 @@
+
+    @Override
+    public String getMessage() {
+        StringBuilder message = new StringBuilder(validationErrors.get(0).getDefaultMessage());
```
можно и так, но лично я предпочитаю просто отдать errors в эксепшен, а форматировать уже специальным классом где понятно какой формат нужен, допустим в json из хэндлера

AndreikaInactive commented on core/src/main/java/com/es/core/cart/MiniCart.java:
```diff
@@ -0,0 +1,35 @@
+    private Map<String, Object> miniCart;
+
+    public MiniCart() {
+        miniCart = new HashMap<>();
```
а почему просто пропертями не засетать в бин?

AndreikaInactive commented on core/src/main/java/com/es/core/config/PropertySourceConfigurer.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        MutablePropertySources envPropSources = ((ConfigurableEnvironment) environment).getPropertySources();
```
а зачем это нужно? коробки должно хватать

VladislawL replied:
Да должно было. Но у меня количество телефонов на страничке находится в пропертис файле, который лежит в вебе. И уже через @Value либо Environment не получается получить это проперти. Как я понял <context:property-placeholder> не сетает проперти в Environment и в других модулях их не видно. По этому я решил в вебе засетать проперти которые связаны с вебом в environment.

VladislawL replied:
А, ещё сорт филды через environment получать во время выполнения + может мне как и стасу проперти phonesOnPage вынести из веба и поместить в application.properties кор модуля? Но филды для сортировки думаю всё равно надо через environment. Может быть есть что то получше?

AndreikaInactive commented on core/src/main/java/com/es/core/dao/JdbcPhoneDao.java:
```diff
@@ -115,22 +134,30 @@ private void add(final Phone phone) {
     }
 
+    private Map<String, Object> getWordsMapFromSearchQuery(String searchQuery) {
+        Map<String, Object> result = new HashMap<>();
```
не совсем понял - зачем нужно это отдельно парсить? можешь привести пример?
мб просто передавать лист

VladislawL replied:
Ну вообще я не вижу лучшего способа как это можно сделать. В строке может быть несколько слов, и я хотел чтобы поиск производился по каждому слову. То есть если строка состоит из 3 слов то создастся мапа с ключами word0, word1, word2, дальше я генерирую строку lower(brand) like :word0 or lower(model) like :word0 or и тд только с word1, word2 для sql запросу.

AndreikaInactive commented on core/src/main/java/com/es/core/services/DefaultPhoneService.java:
```diff
@@ -2,33 +2,53 @@
+
+    @Autowired
+    public DefaultPhoneService(Environment environment) {
+        phonesOnPage = Integer.parseInt(environment.getProperty("phonesOnPage"));
```
можно инжектить через ${} в хмл или @Value

AndreikaInactive commented on core/src/main/java/com/es/core/validators/QuantityValidator.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class QuantityValidator {
```
если для ui - имплементь Validator спринговый

VladislawL replied:
У меня так было до того как открыл пулреквест. Вроде порешали на том что валидность quantity буду проверять перед добавлением

AndreikaInactive commented on core/src/main/resources/config/sortFields.properties:
```diff
@@ -0,0 +1,4 @@
+brand=brand
```
можно в обычный пропертис запихнуть, с каким префиксом

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/config/WebAppConfig.java:
```diff
@@ -0,0 +1,21 @@
+    @Bean
+    @Scope(value = "request", proxyMode = ScopedProxyMode.TARGET_CLASS)
+    @AjaxErrorMessageQualifier
+    public Map<String, Object> errorMessage() {
```
прикольно, а зачем это?

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/AjaxCartController.java:
```diff
@@ -23,18 +24,21 @@
+
+            miniCart.setSubTotalPrice(cart.getSubTotalPrice());
+            miniCart.setItemsNumber(cart.getCartItems().size());
+            return ResponseEntity.ok(miniCart.getMiniCart());
```
сразу бин отдавай ) странно читается

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/advicer/AjaxExceptionHandler.java:
```diff
@@ -1,20 +1,31 @@
 
+    @Autowired
+    @AjaxErrorMessageQualifier
+    private Map<String, Object> errorMessage;
```
так а если просто тут маппить exception в какой то простой error формат? типо список бинов - название ошибки и  месседж с пояснением

AndreikaInactive commented on web/src/main/webapp/WEB-INF/applicationContext.xml:
```diff
@@ -5,4 +5,20 @@
+                <bean class="org.springframework.core.io.support.ResourcePropertySource">
+                    <constructor-arg value="classpath:config/sortFields.properties" />
+                </bean>
+                <bean class="org.springframework.core.io.support.ResourcePropertySource">
```
вроде можно в коробочный property xml tag запихнуть доп сорсы, лучше без кастомного бина

BuisnessMonkeyy committed:
Fixed addPhone and updatePhone in HttpSessionService
Replaced PropertySourceConfigurer with PropertiesFactoryBean
Added PriceFormatter, SortFieldPropertyService

BuisnessMonkeyy committed:
Deleted PageData

AndreikaInactive commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -34,45 +32,41 @@ public Cart getCart() {
-            add(phoneId, quantity);
+            addNewCartItem(phoneId, quantity);
         }
+        cart.setSubTotalPrice(priceCalculatorService.calculateSubtotalPrice(cart));
```
прайсы на карте могут быть разными, лучше все их сетать в методе прайс сервиса

AndreikaInactive commented on core/src/main/java/com/es/core/cart/MiniCart.java:
```diff
@@ -5,31 +5,37 @@
-        miniCart = new HashMap<>();
-        miniCart.put("itemsNumber", 0);
-        miniCart.put("subTotalPrice", BigDecimal.ZERO);
+        subTotalPrice = BigDecimal.ZERO.toString();
```
юзай бд, а на фронте уже форматировать в нужное представление, универсальнее и например формат по локали можно юзать
если стринг, то уже ничего с ним не сделаешь

AndreikaInactive commented on core/src/main/java/com/es/core/dao/JdbcPhoneDao.java:
```diff
@@ -30,7 +30,7 @@
 
     @Autowired
-    private Environment environment;
+    private PropertyService sortFieldPropertyService;
```
как и брату - квери через лайки и вайлдарды будет проще, проперти сервис не нужен - всё можно инжектить через спринг конвершен
посмотри какие примеры для работы с property placeholder

VladislawL replied:
Через ? квери как-то наоборот сложнее получается, т.к. там идёт brand like or model like. Поэтому слова нужно будет продублировать. Как мне кажется NamedJdbcTemplate в этом плане удобнее, т.к. параметр я могу переиспользовать. Можно попробовать поменять стратегию поиска, например, делать поиск не по каждому слову а на все searchQuery

AndreikaInactive commented on core/src/main/java/com/es/core/services/DefaultPhoneService.java:
```diff
@@ -18,13 +17,9 @@
     @Resource
     private PhoneDao phoneDao;
 
+    @Value("#{propertySource['phonesOnPage']}")
```
просто по ключу инжекть

VladislawL replied:
через ${} не хочет работать. Падает с ошибкой Could not resolve placeholder 'phonesOnPage' in value "${phonesOnPage}". У меня же phonesOnPage лежит в вебе + я использую PropertiesFactoryBean. Наверное из-за этого так не работает, у Стаса просто все проперти лежат в application.properties кор модуля и всё окей,. Может так же сделать?

AndreikaInactive commented on core/src/test/java/com/es/core/services/DefaultPhoneServiceTest.java:
```diff
@@ -21,7 +21,7 @@
     private DefaultPhoneService phoneService;
 
-    @Value("${phonesOnPage}")
+    @Value("#{propertySource['phonesOnPage']}")
```
и тут

AndreikaInactive commented on web/src/main/java/com/es/phoneshop/web/controller/AjaxCartController.java:
```diff
@@ -27,16 +27,19 @@
             Cart cart = cartService.getCart();
 
-            miniCart.setSubTotalPrice(cart.getSubTotalPrice());
+            miniCart.setSubTotalPrice(priceFormatter.format(cart.getSubTotalPrice(), LocaleContextHolder.getLocale()));
```
сделай статический метод в миникарте fromCart, будет красивее
форматировать можно и на фронте

BuisnessMonkeyy committed:
Added Attribute and AttributeService, fromCart method in MiniCart
Fixed PriceCalculator
Added defaultCurrency bean
Deleted SortFieldPropertySource
